lib/compiler/src/beam_block.erl

```diff
@@ -33,15 +33,50 @@ module({Mod,Exp,Attr,Fs0,Lc}, _Opts) ->
 
 function({function,Name,Arity,CLabel,Is0}) ->
     try
-        Is1 = blockify(Is0),
-        Is = embed_lines(Is1),
+        Is1 = swap_opt(Is0),
+        Is2 = blockify(Is1),
+        Is = embed_lines(Is2),
         {function,Name,Arity,CLabel,Is}
     catch
         Class:Error:Stack ->
 	    io:fwrite("Function: ~w/~w\n", [Name,Arity]),
 	    erlang:raise(Class, Error, Stack)
     end.
 
+%%%
+%%% Try to use a `swap` instruction instead of a sequence of moves.
+%%%
+%%% Note that beam_ssa_codegen generates `swap` instructions only for
+%%% the moves within a single SSA instruction (such as `call`), not
+%%% for the moves generated by a sequence of SSA instructions.
+%%% Therefore, this optimization is needed.
+%%%
+
+swap_opt([{move,Reg1,{x,X}=Temp}=Move1,
+          {move,Reg2,Reg1}=Move2,
+          {move,Temp,Reg2}=Move3|Is]) when Reg1 =/= Temp ->
+    case is_unused(X, Is) of
+        true ->
+            [{swap_temp,Reg1,Reg2,Temp}|swap_opt(Is)];
+        false ->
+            [Move1|swap_opt([Move2,Move3|Is])]
+    end;
+swap_opt([I|Is]) ->
+    [I|swap_opt(Is)];
+swap_opt([]) -> [].
+
+is_unused(X, [{call,A,_}|_]) when A =< X -> true;
+is_unused(X, [{call_ext,A,_}|_]) when A =< X -> true;
+is_unused(X, [{make_fun2,_,_,_,A}|_]) when A =< X -> true;
+is_unused(X, [{move,Src,Dst}|Is]) ->
```

Comment:
The loader code had some checks for optimising moves to swaps before a `send` instruction. Is there a reason this is no longer done?

Reply:
Most of the `swap` instructions are generated by `beam_ssa_codegen`, and it will handle `send` (it is still a call at that point).

In `beam_block` I only handled the instructions that would make a difference that would be seen in the modules compiled by `scripts/diffable`.